File: app/src/main/java/com/controlmoblie/asr/SpeechRecognizerManager.java
package com.controlmoblie.asr;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;

import com.k2fsa.sherpa.onnx.EndpointConfig;
import com.k2fsa.sherpa.onnx.EndpointRule;
import com.k2fsa.sherpa.onnx.FeatureConfig;
import com.k2fsa.sherpa.onnx.HomophoneReplacerConfig;
import com.k2fsa.sherpa.onnx.OnlineCtcFstDecoderConfig;
import com.k2fsa.sherpa.onnx.OnlineLMConfig;
import com.k2fsa.sherpa.onnx.OnlineModelConfig;
import com.k2fsa.sherpa.onnx.OnlineRecognizer;
import com.k2fsa.sherpa.onnx.OnlineRecognizerConfig;
import com.k2fsa.sherpa.onnx.OnlineStream;
import com.k2fsa.sherpa.onnx.OnlineTransducerModelConfig;

import java.io.File;

public class SpeechRecognizerManager {
    private static final String TAG = "SpeechRecognizerManager";
    private static final int SAMPLE_RATE = 16000;

    public interface RecognitionListener {
        void onPartialResult(String text);
        void onFinalResult(String text);
        void onError(String message);
        void onReady();
    }

    private final String modelPath;
    private OnlineRecognizer recognizer;
    private OnlineStream stream;
    private AudioRecord audioRecord;
    private Thread captureThread;
    private volatile boolean isListening = false;
    private RecognitionListener listener;

    public SpeechRecognizerManager(String modelPath) {
        this.modelPath = modelPath;
    }

    public void setListener(RecognitionListener listener) {
        this.listener = listener;
    }

    public boolean init() {
        try {
            File dir = new File(modelPath);
            File[] allFiles = dir.listFiles();
            java.util.List<File> onnxFiles = new java.util.ArrayList<>();
            if (allFiles != null) {
                collectOnnxFiles(dir, onnxFiles);
            }
            if (onnxFiles.isEmpty()) {
                Log.e(TAG, "No .onnx files found in " + modelPath);
                if (listener != null) listener.onError("模型文件缺失");
                return false;
            }
            String encoder = findFileByName(onnxFiles, "encoder");
            String decoder = findFileByName(onnxFiles, "decoder");
            String joiner = findFileByName(onnxFiles, "joiner");
            Log.d(TAG, "encoder=" + encoder + ", decoder=" + decoder + ", joiner=" + joiner);

            String tokensFile = findFileByName(dir, "tokens.txt");
            if (tokensFile == null) tokensFile = modelPath + "/tokens.txt";

            FeatureConfig featConfig = new FeatureConfig(SAMPLE_RATE, 80, 0.0f);

            OnlineTransducerModelConfig transducer = new OnlineTransducerModelConfig(encoder, decoder, joiner);
            OnlineModelConfig modelConfig = new OnlineModelConfig();
            modelConfig.setTransducer(transducer);
            modelConfig.setTokens(tokensFile);
            modelConfig.setNumThreads(2);
            modelConfig.setProvider("cpu");
            modelConfig.setDebug(false);

            EndpointRule rule1 = new EndpointRule(false, 2.4f, 0.0f);
            EndpointRule rule2 = new EndpointRule(true, 1.2f, 0.0f);
            EndpointRule rule3 = new EndpointRule(false, 0.0f, 20.0f);
            EndpointConfig endpointConfig = new EndpointConfig(rule1, rule2, rule3);

            OnlineRecognizerConfig config = new OnlineRecognizerConfig(
                featConfig,
                modelConfig,
                new OnlineLMConfig(),
                new OnlineCtcFstDecoderConfig(),
                new HomophoneReplacerConfig(),
                endpointConfig,
                true,
                "greedy_search",
                4,
                "",
                1.5f,
                "",
                "",
                0.0f
            );

            recognizer = new OnlineRecognizer(null, config);
            Log.d(TAG, "OnlineRecognizer initialized");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Failed to load model", e);
            if (listener != null) listener.onError("模型加载失败: " + e.getMessage());
            return false;
        }
    }

    private void collectOnnxFiles(File dir, java.util.List<File> result) {
        File[] files = dir.listFiles();
        if (files == null) return;
        for (File f : files) {
            if (f.isDirectory()) {
                collectOnnxFiles(f, result);
            } else if (f.getName().endsWith(".onnx")) {
                result.add(f);
            }
        }
    }

    private String findFileByName(java.util.List<File> files, String keyword) {
        for (File f : files) {
            if (f.getName().toLowerCase().contains(keyword.toLowerCase())) {
                return f.getAbsolutePath();
            }
        }
        return files.get(0).getAbsolutePath();
    }

    private String findFileByName(File dir, String name) {
        File[] files = dir.listFiles();
        if (files == null) return null;
        for (File f : files) {
            if (f.isDirectory()) {
                String found = findFileByName(f, name);
                if (found != null) return found;
            } else if (f.getName().equals(name)) {
                return f.getAbsolutePath();
            }
        }
        return null;
    }

    public void startListening() {
        if (isListening) stopListening();
        OnlineRecognizer rec = recognizer;
        if (rec == null) {
            if (listener != null) listener.onError("模型未加载");
            return;
        }

        try {
            stream = rec.createStream("");

            int minBufSize = AudioRecord.getMinBufferSize(
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
            );
            int bufferSize = Math.max(minBufSize, SAMPLE_RATE / 5);

            audioRecord = new AudioRecord(
                MediaRecorder.AudioSource.MIC,
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize
            );

            if (audioRecord.getState() != AudioRecord.STATE_INITIALIZED) {
                if (listener != null) listener.onError("麦克风初始化失败");
                return;
            }

            audioRecord.startRecording();
            isListening = true;
            if (listener != null) listener.onReady();

            captureThread = new Thread(() -> {
                short[] buffer = new short[bufferSize / 2];
                String lastPartialText = "";

                while (isListening) {
                    int len = audioRecord.read(buffer, 0, buffer.length);
                    if (len <= 0) continue;

                    float[] samples = new float[len];
                    for (int i = 0; i < len; i++) {
                        samples[i] = buffer[i] / 32768f;
                    }
                    OnlineStream s = stream;
                    if (s == null) break;
                    s.acceptWaveform(samples, SAMPLE_RATE);

                    while (rec.isReady(s)) {
                        rec.decode(s);
                    }

                    String text = rec.getResult(s).getText().trim();

                    if (!text.isEmpty() && rec.isEndpoint(s)) {
                        rec.reset(s);
                        if (listener != null) listener.onFinalResult(text);
                        lastPartialText = "";
                    } else if (!text.isEmpty() && !text.equals(lastPartialText)) {
                        if (listener != null) listener.onPartialResult(text);
                        lastPartialText = text;
                    }
                }
            });
            captureThread.setName("Asr-capture");
            captureThread.setPriority(Thread.MAX_PRIORITY);
            captureThread.start();

        } catch (Exception e) {
            Log.e(TAG, "Failed to start listening", e);
            if (listener != null) listener.onError("启动识别失败: " + e.getMessage());
        }
    }

    public void stopListening() {
        isListening = false;
        if (captureThread != null) {
            try {
                captureThread.join(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            captureThread = null;
        }
        if (audioRecord != null) {
            audioRecord.stop();
            audioRecord.release();
            audioRecord = null;
        }
        if (stream != null) {
            stream.release();
            stream = null;
        }
    }

    public void release() {
        stopListening();
        if (recognizer != null) {
            recognizer.release();
            recognizer = null;
        }
    }
}

File: app/src/main/java/com/controlmoblie/asr/SpeechRecognizerManager.kt
package com.controlmoblie.asr

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.util.Log
import com.k2fsa.sherpa.onnx.OnlineTransducerModelConfig
import com.k2fsa.sherpa.onnx.EndpointConfig
import com.k2fsa.sherpa.onnx.EndpointRule
import com.k2fsa.sherpa.onnx.FeatureConfig
import com.k2fsa.sherpa.onnx.HomophoneReplacerConfig
import com.k2fsa.sherpa.onnx.OnlineCtcFstDecoderConfig
import com.k2fsa.sherpa.onnx.OnlineLMConfig
import com.k2fsa.sherpa.onnx.OnlineModelConfig
import com.k2fsa.sherpa.onnx.OnlineRecognizer
import com.k2fsa.sherpa.onnx.OnlineRecognizerConfig
import com.k2fsa.sherpa.onnx.OnlineStream
import java.io.File
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow

sealed class AsrEvent {
    data class PartialResult(val text: String) : AsrEvent()
    data class FinalResult(val text: String) : AsrEvent()
    data class Error(val message: String) : AsrEvent()
    object Ready : AsrEvent()
}

class SpeechRecognizerManager(private val modelPath: String) {

    private var recognizer: OnlineRecognizer? = null
    private var stream: OnlineStream? = null
    private var audioRecord: AudioRecord? = null
    private var captureThread: Thread? = null
    private val _events = Channel<AsrEvent>(Channel.CONFLATED)
    val events: Flow<AsrEvent> = _events.receiveAsFlow()
    private var isListening = false

    fun init(): Boolean {
        return try {
            val dir = File(modelPath)
            val onnxFiles = dir.listFiles { f -> f.name.endsWith(".onnx") }
            if (onnxFiles.isNullOrEmpty()) {
                Log.e(TAG, "No .onnx files found in $modelPath")
                _events.trySend(AsrEvent.Error("模型文件缺失"))
                return false
            }
            val encoder = onnxFiles.find { it.name.contains("encoder", ignoreCase = true) }?.absolutePath
                ?: onnxFiles.first().absolutePath
            val decoder = onnxFiles.find { it.name.contains("decoder", ignoreCase = true) }?.absolutePath
                ?: onnxFiles.first().absolutePath
            val joiner = onnxFiles.find { it.name.contains("joiner", ignoreCase = true) }?.absolutePath
                ?: onnxFiles.first().absolutePath
            Log.d(TAG, "encoder=$encoder, decoder=$decoder, joiner=$joiner")

            val featConfig = FeatureConfig(SAMPLE_RATE, 80, 0.0f)

            val modelConfig = OnlineModelConfig().apply {
                transducer = OnlineTransducerModelConfig(encoder, decoder, joiner)
                tokens = "$modelPath/tokens.txt"
                numThreads = 2
                provider = "cpu"
                debug = false
            }

            val rule1 = EndpointRule(false, 2.4f, 0.0f)
            val rule2 = EndpointRule(true, 1.2f, 0.0f)
            val rule3 = EndpointRule(false, 0.0f, 20.0f)
            val endpointConfig = EndpointConfig(rule1, rule2, rule3)

            val config = OnlineRecognizerConfig(
                featConfig = featConfig,
                modelConfig = modelConfig,
                lmConfig = OnlineLMConfig(),
                ctcFstDecoderConfig = OnlineCtcFstDecoderConfig(),
                hr = HomophoneReplacerConfig(),
                endpointConfig = endpointConfig,
                enableEndpoint = true,
                decodingMethod = "greedy_search",
                maxActivePaths = 4,
                hotwordsFile = "",
                hotwordsScore = 1.5f,
                ruleFsts = "",
                ruleFars = "",
                blankPenalty = 0.0f
            )

            recognizer = OnlineRecognizer(config = config)
            Log.d(TAG, "OnlineRecognizer initialized")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load model", e)
            _events.trySend(AsrEvent.Error("模型加载失败: ${e.message}"))
            false
        }
    }

    fun startListening() {
        if (isListening) stopListening()
        val rec = recognizer ?: run {
            _events.trySend(AsrEvent.Error("模型未加载"))
            return
        }

        try {
            stream = rec.createStream()

            val minBufSize = AudioRecord.getMinBufferSize(
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
            )
            val bufferSize = maxOf(minBufSize, SAMPLE_RATE / 5)

            audioRecord = AudioRecord(
                MediaRecorder.AudioSource.MIC,
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize
            )

            if (audioRecord?.state != AudioRecord.STATE_INITIALIZED) {
                _events.trySend(AsrEvent.Error("麦克风初始化失败"))
                return
            }

            audioRecord?.startRecording()
            isListening = true
            _events.trySend(AsrEvent.Ready)

            captureThread = Thread {
                val buffer = ShortArray(bufferSize / 2)
                var lastPartialText = ""

                while (isListening) {
                    val len = audioRecord?.read(buffer, 0, buffer.size) ?: -1
                    if (len <= 0) continue

                    val samples = FloatArray(len) { buffer[it] / 32768f }
                    val s = stream ?: break
                    s.acceptWaveform(samples, SAMPLE_RATE)

                    while (rec.isReady(s)) {
                        rec.decode(s)
                    }

                    val result = rec.getResult(s)
                    val text = result.text.trim()

                    if (text.isNotEmpty() && rec.isEndpoint(s)) {
                        rec.reset(s)
                        _events.trySend(AsrEvent.FinalResult(text))
                        lastPartialText = ""
                    } else if (text.isNotEmpty() && text != lastPartialText) {
                        _events.trySend(AsrEvent.PartialResult(text))
                        lastPartialText = text
                    }
                }
            }.apply {
                name = "Asr-capture"
                priority = Thread.MAX_PRIORITY
                start()
            }

        } catch (e: Exception) {
            Log.e(TAG, "Failed to start listening", e)
            _events.trySend(AsrEvent.Error("启动识别失败: ${e.message}"))
        }
    }

    fun stopListening() {
        isListening = false
        captureThread?.join(500)
        captureThread = null
        audioRecord?.apply {
            stop()
            release()
        }
        audioRecord = null
        stream?.release()
        stream = null
    }

    fun release() {
        stopListening()
        recognizer?.release()
        recognizer = null
    }

    companion object {
        private const val TAG = "SpeechRecognizerManager"
        private const val SAMPLE_RATE = 16000
    }
}
